07-errors.md

```diff
@@ -0,0 +1,311 @@
+| 11-15 | reserved                   |                                                            |
+
+Notes:
+
+1. If a message sender doesn't exist on-chain, the VM will return a receipt with the
+   `ExitCode::SysErrSenderInvalid` exit code.
+2. If an actor calls `vm::abort` with a system exit code, that exit code will be replaced with
+   `ExitCode::SysErrIllegalExitCode`.
+3. If an actor panics/crashes/etc., the exit code `ExitCode::SysErrIllegalInstruction` will be
+   returned.
+4. If a message runs out of gas, the VM will return a receipt with the `ExitCode::SysErrOutOfGas`
+   exit code.
+5. Most of these exit codes are only reflected on-chain and will not be observed in internal sends.
+   The exceptions are:
+    - `SysErrIllegalExitCode` if the called actor aborted with a system exit code.
+    - `SysErrIllegalInstruction` if the called actor paniced.
+
+Changes from pre-FVM Filecoin:
+
+1. Removed `ExitCode::SysErrInvalidMethod` (3). Dispatch is now handled inside actor code.
+2. Removed `ExitCode::SysErrForbidden` (8). Caller verification is a userspace operation, we'll now
+   exit with `ExitCode::ErrForbidden`.
+4. Replaced `ExitCode::SysErrIllegalArgument` with `ExitCode::SysErrAssertionFailed` to more
+   accurately reflect its usage. When this appears on-chain, it means that a message hit some
+   internal Filecoin assert that shouldn't happen.
+
+### Standard Non-Zero Exit  Codes
+
+Standard non-zero exit codes are codes returned by actors to indicate common failure conditions.
+
+| Code  | Status                 | Description                                         |
+|-------|------------------------|-----------------------------------------------------|
+| 16    | `ErrIllegalArgument`   | invalid message parameters                          |
+| 17    | `ErrNotFound`          | message referenced something that doesn't exist     |
+| 18    | `ErrForbidden`         | operation forbidden                                 |
+| 19    | `ErrInsufficientFunds` | insufficient funds for operation                    |
+| 20    | `ErrIllegalState`      | the actor is in an illegal state                    |
+| 21    | `ErrSerialization`     | actor failed to serialize or deserialize some state |
+| 22    | `ErrUnhandledMessage`  | actor cannot handle this message                    |
+| 23    | `ErrUnspecified`       | the actor failed with an unspecified error          |
+| 24-31 | reserved               |                                                     |
+
+Changes from pre-FVM Filecoin:
+
+1. Added `ExitCode::ErrUnhnadledMessage` (replacing `ExitCode::SysErrInvalidMethod`).
+2. Added `ExitCode::ErrUnspecified`. This generally means "there's a bug and I have no idea what
+   went wrong". The alternative would be for the actor to _panic_ (leading to an
+   `ExitCode::ErrIllegalInstruction` syscall error), but it may be useful to distinguish between
+   "crashing" and "something unknown went wrong".
```

Comment:
nit: The usage of "may" introduces non-determinism here. Since this was a bit of a controversial point, I suggest we either: (a) fully norm this out, or (b) we explicitly mention that more details will follow in a future iteration.

Reply:
This is fully specified as-is. That "may" isn't saying anything about the system, it's saying that the _actor_ may want to distinguish between these two cases and that the system allows for that.